src/Common/GuardedPoolAllocator.h

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include <Common/GuardedPoolAllocatorCommon.h>
+#include <Common/GuardedPoolAllocatorOptions.h>
+
+#include <mutex>
+
+#include <cinttypes>
+
+namespace clickhouse_gwp_asan
+{
+
+class GuardedPoolAllocator;
+
+GuardedPoolAllocator * getAllocator();
+
+class GuardedPoolAllocator
+{
+public:
+    constexpr GuardedPoolAllocator() { }
+    GuardedPoolAllocator(const GuardedPoolAllocator &) = delete;
+    GuardedPoolAllocator & operator=(const GuardedPoolAllocator &) = delete;
+
+    ~GuardedPoolAllocator() = default;
+
+    void init(const clickhouse_gwp_asan::options::Options & options);
+
+    /// Lock pool_reservation_mutex to prevent any in-progress allocation.
+    /// Intended to be used in signal handler on crash to prevent any possible
+    /// metadata corruption when trying to diagnose error.
+    void lock();
+
+    void * allocate(size_t size, size_t alignment);
+    void deallocate(void * ptr);
+
+    inline ALWAYS_INLINE bool shouldSample()
+    {
+        // next_sample_counter == 0 means we "should regenerate the counter".
+        //                   == 1 means we "should sample this allocation".
+        // AdjustedSampleRatePlusOne is designed to intentionally underflow. This
```

Comment:
Comment from some other place? Let's put a comment on which code this is based on (generally in each header)